services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedRegistrationService.java

```diff
@@ -0,0 +1,369 @@
+import org.eclipse.hono.deviceregistry.util.DeviceRegistryUtils;
+import org.eclipse.hono.deviceregistry.util.Versioned;
+import org.eclipse.hono.service.management.Id;
+import org.eclipse.hono.service.management.OperationResult;
+import org.eclipse.hono.service.management.Result;
+import org.eclipse.hono.service.management.device.Device;
+import org.eclipse.hono.service.management.device.DeviceManagementService;
+import org.eclipse.hono.service.registration.AbstractRegistrationService;
+import org.eclipse.hono.service.registration.RegistrationService;
+import org.eclipse.hono.tracing.TracingHelper;
+import org.eclipse.hono.util.RegistrationConstants;
+import org.eclipse.hono.util.RegistrationResult;
+import org.eclipse.hono.util.RegistryManagementConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.stereotype.Component;
+
+import com.mongodb.ErrorCategory;
+import com.mongodb.MongoException;
+
+import io.opentracing.Span;
+import io.vertx.core.AbstractVerticle;
+import io.vertx.core.Future;
+import io.vertx.core.Promise;
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.mongo.IndexOptions;
+import io.vertx.ext.mongo.MongoClient;
+import io.vertx.ext.mongo.MongoClientDeleteResult;
+import io.vertx.ext.mongo.MongoClientUpdateResult;
+
+/**
+ * This is an implementation of the device registration service and the device management service where data 
+ * is stored in a mongodb database.
+ *
+ * @see <a href="https://www.eclipse.org/hono/docs/api/device-registration/">Device Registration API</a>
+ * @see <a href="https://www.eclipse.org/hono/docs/api/management/">Device Registry Management API</a>
+ */
+@Component
+@Qualifier("serviceImpl")
+@ConditionalOnProperty(name = "hono.app.type", havingValue = "mongodb", matchIfMissing = true)
+public class MongoDbBasedRegistrationService extends AbstractVerticle
+        implements DeviceManagementService, RegistrationService {
+
+    private static final Logger log = LoggerFactory.getLogger(MongoDbBasedRegistrationService.class);
+    private MongoClient mongoClient;
+    private MongoDbBasedRegistrationConfigProperties config;
+    private MongoDbCallExecutor mongoDbCallExecutor;
+
+    /**
+     * Registration service, based on {@link AbstractRegistrationService}.
+     * <p>
+     * This helps work around Java's inability to inherit from multiple base classes. We create a new Registration
+     * service, overriding the implementation of {@link AbstractRegistrationService} with the implementation of our
+     * {@link MongoDbBasedRegistrationService#getDevice(String, String, Span)}.
+     */
+    private final AbstractRegistrationService registrationService = new AbstractRegistrationService() {
```

Comment:
this will be very hard to test. why don't you simply extend *AbstractRegistrationService* and also implement `Verticle`?